src/layoutBuilder.js

```diff
@@ -388,75 +495,224 @@ function convertWordsToCodepoints(lineAsArrayOfCodepoints) {
+	// individual RTL words as they will be reversed at the character
+	// level.
+
+	// Contains the codepoints for each word in the line.
+	var arrayOfCodePoints = convertWordsToCodepoints(bidiString.string_arr);
+
+	// Contains the strings for each word in the line.
+	var arrayOfTransformedWords = [];
+
+	// With the line converted to an array of words, with the words represented as
+	// codepoints, the words must be individually re-run through the BIDI algorithm
+	// to reverse the characters in each RTL (as the initial BIDI process reverses each
+	// character in the RTL words).
+
+	// Pass the codepoints for each word through String.fromCharCode
+	for (
+		var groupIndex = 0;
+		groupIndex < arrayOfCodePoints.length;
+		groupIndex++
+	) {
+		// Get current word as a string
+		var groupString = spreadify(
+			String.fromCharCode,
+			String
+		)(arrayOfCodePoints[groupIndex]);
+		// Run the word through BIDI to reorder the characters
+		var bidiWord = bidi.from_string(groupString, { direction: 'RTL' });
+		bidiWord.reorder_visually();
+
+		// Push the word to what will be the final line array
+		arrayOfTransformedWords.push(bidiWord.toString());
+	}
+
+	// Pass the transformed words, as a single string, through the buildInlines utility.
+	// Include original styling of the node as a whole e.g. if the whole body of text is bold
+	// then make sure this gets passed through.
+	var updatedInlines = textTools.buildInlines(
+		[{ text: arrayOfTransformedWords.join(''), style: textNode.style }],
+		styleStack
+	);
+
+	// Wipe the existing words (inlines) from this line ...
+	line.inlines = [];
+	// ... and replace them with our new BIDI-ified, reversed inlines
+	updatedInlines.items.forEach(function (inline) {
+		// Where the BIDI algorithm has appropriately transformed the content
+		// we can be confident that the postions of the words have been reversed.
+		// For example, a word that was at position 0 prior to the transformation will
+		// now likely be at the end of the array. This gives us a way of extracting the
+		// styling for each word before it was transformed and apply it to the
+		// transformed word.
+
+		// TODO I don't know how this will stack up again LTR words. Will need
+		// to evaluate once we have mixed fonts support.
+
+		// This needs to use find because the arrays inlinesBeforeTransformation and
+		// updatedInlines are not necessarily the same lengths (due to how the reversal
+		// interacts with whitespace and punctuation in some cases)
+		let oldInlineIndex = inlinesBeforeTransformation.find(
+			(oldInline) => inline.text.trim() === oldInline.text.trim()
+		);
+		if (oldInlineIndex === -1) {
+			oldInlineIndex = inlinesBeforeTransformation.find((oldInline) =>
+				inline.text.includes(oldInline.text.trim())
+			);
+		}
+		if (oldInlineIndex === -1) {
+			oldInlineIndex = inlinesBeforeTransformation.find((oldInline) =>
+				oldInline.text.includes(inline.text.trim())
+			);
+		}
```

Comment:
This is the only amendment I have made to the old `transformLineForRtl` function.

The assumption that two array lengths are identical is not valid and this results in highlighting issues in the documents. I've tried to maximise the chances of finding the previous inline (so allowing styling to be correctly applied) but this may be pretty slow.

Reply:
3 loops does feel excessive, especially the last 2.

I see you're looking for whether the oldInline value is a subset of the current and vice versa - does the order matter? I wonder if we could do a single for loop and terminate if any of the conditions pass

A unit test for a function to get `oldInlineIndex` would help here if we do decide to simplify it.